### ISSUE TITLE
[Feature request] Per-panel variables

### ISSUE THREAD

**Issue by zenmetsu** [grafana/grafana#8408]
Due to the limitation of not being able to utilize template variables with alerting, it would be very useful to allow for variable declaration within individual panels.  This would allow an admin to duplicate a set of panels which reference a $hostname, for instance, and change the variable assignment once within each individual panel to automatically change title, queries, etc.  

This would be preferable to a hack to allow template variables to work with alerting since it prevents the possibility of a user from removing hosts from a template selection and thus disabling alerting.

**Comment by belm0** [comment]
Noting significant use cases from a separately filled issue:

**Why is this needed:**

For example, a variable to chose the displayed quantile, or N of a topk query, or a filter applied to the metric's labels.

Dashboard variables aren't useful for this, because you'd like to set each panel independently, and it would result in an unmanageable and confusing number of similarly-named variables.

**Comment by belm0** [comment]
I'm not familiar with grafana internals, but here's what I imagine needs to be done:

  * [ ] amend JSON model schema - items in `panels` will need a `templating` attribute similar to what exits globally to the dashboard
  * [ ] change variable lookup logic to check panel scope first
  * [ ] represent panel variables in page URL - the panel variable names need to be qualified by panel ID number
  * [ ] add panel window UI to access panel variables
  * [ ] add variable tab to panel config UI

These are roughly in the order they should be implemented.   Proof of concept can be demonstrated by implementing only the first 2 items.

**Comment by adamf663** [comment]
This really needs to be implemented.  Grafana's variables are too primitive.  I'm probably going to have to dump grafana for an excel spreadsheet for this reason.

**Comment by belm0** [comment]
imagine if there were per-panel variables

you could use a variable to change the query function of each panel...

![grafana_variable_func](https://user-images.githubusercontent.com/1708631/101599838-c48e2c80-3a3d-11eb-9eda-379b27784e77.gif)

**Comment by mukteshkrmishra** [comment]
This will be really cool.

**Comment by AntiTenzor** [comment]
Panel variables will be very useful in complex dynamic dashboards.
For example, dashboard-level variable may be used to select some "Categories".
After that at a Panel level we could define panel variable of items in this "dashboard-level" categories.

To make things more straightforward such "panel variables" may use special prefix to distinguish them from "dashboard vars".

**Comment by nicolaerosia** [comment]
this would be great!

**Comment by edengabay** [comment]
I would love to have such feature. Thanks.

**Comment by JaimeAvalosIDA** [comment]
I think this feature would big a huge new set in the interactivity of grafana dashboards. Thank you for all your work.

**Comment by zenglime** [comment]
agree, I need this feature also them I can merge multiple dashboards into one dashboard with panel-level vairables

**Comment by blukhand** [comment]
agree, the pannel setting would be much more flexible with this feature

**Comment by PWSys** [comment]
This would be great for my use case.

**Comment by Abuu001** [comment]
I agree that this feature should be added.  since I have several panels in one dashboard and some variables are dependent only in one panel. so it will be great if there is an option to set a variable to appear to specific panel.

**Comment by WoZ** [comment]
The same for me. +1

**Comment by sweller999** [comment]
Agree with everyone above. There are some situations where dashboard-level variables simply do not provide the granularity needed to convey information in the most efficient way. Yes, there are workarounds, but the result is a cluttered dash which requires grokking to understand vs. something I can just send my boss that he can use right away

**Comment by adamf663** [comment]
five years later...

**Comment by clarencedesmarais** [comment]
We need this <3

**Comment by dwyer125** [comment]
+1 please,

**Comment by abilash-sethu** [comment]
I was looking for it, This would have been great <3

**Comment by missonganesh** [comment]
+1 - Would be very useful!

**Comment by FoodMuncher** [comment]
Please could this be added!

**Comment by plysytsya** [comment]
Would be great to have that

**Comment by hzwwww** [comment]
It's a tremendous convenience in some situations, please considering that feature🙏

**Comment by SamReece** [comment]
This feature would be a very helpful update to panels on my current dashboards.

**Comment by lavalex** [comment]
Just was looking for something like this, very needed.

**Comment by itsnull12** [comment]
+1 very very needed

**Comment by gabivlj** [comment]
+1, this would be useful

**Comment by ns-kkodiyan** [comment]
+1 , got here looking for this feature

**Comment by avinashkris9** [comment]
+1

**Comment by rishabhverma00** [comment]
Yes, I am also looking for this feature.

**Comment by ryaudio** [comment]
+1

**Comment by sweller999** [comment]
The year is 3948. Civilizations have risen and fallen. Humanity has transcended its mortal form and has integrated with machines, thus achieving immortality. There is no hunger, war or disease, and mankind travels freely throughout the stars aboard ships capable of FTL travel. 
...
...
yet STILL, Grafana does not have dashboard-level variable support

**Comment by mourya92** [comment]
+1

**Comment by JPS-Northvolt** [comment]
+1

**Comment by Loki087** [comment]
+1

**Comment by AsimMaan** [comment]
Is there anyone who can develop this feature in Grafana? (consider it a paid request)

**Comment by MounikaBanda13** [comment]
+1 much needed

**Comment by serrrios** [comment]
+1

**Comment by adamf663** [comment]
It's been six years;  time to find alternative solutions.  The folks at grafana have made it painfully clear that they don't care in the slightest about the usability of their product.  They're fine with a user hostile UI.

**Comment by arjungarg07** [comment]
Can we escalate this feature request. This is an all time want now.

**Comment by sweller999** [comment]
I just talked to the Grafana people at Scale 20X last week and was told Grafana Scenes is supposed to address this. They are well aware of the issue, and it seems that a lot of the holdup in getting this implemented is the complicated nature of all of the 3rd party apps and plugins they support.

As for regular Grafana, I was told not to hold my breath

**Comment by IerwinDE** [comment]
If we were able to place the variables selection UI anywhere else than the default place would already solve a lot of issues.
I can build my panels to use single variables that exists for the whole dashboard. Beeing able to place the selector close to the panel where the variable actually has an effect would be so useful (and I guess it wouldn't break the whole way variables are implemented)

**Comment by belm0** [comment]
I had outlined the tasks to implement per-panel variables here.  While it's not rocket science, it would be important to find someone at Grafana willing to steward the experiment and review PR's.  I would think someone out there could convince their employer to let them spend time on this.

https://github.com/grafana/grafana/issues/8408#issuecomment-562903857

**Comment by wkneewalden** [comment]
+1

**Comment by MMusicLLC** [comment]
+1

**Comment by Shinekhuu** [comment]
+1

**Comment by Mrkezii** [comment]
+1 PLEASSSEEEEEEEEEEE

**Comment by PartlyFluked** [comment]
+1

**Comment by kevinlinxc** [comment]
Sorely need this feature. It makes it so that laypeople can have an interface to modify a dashboard panel without knowing SQL, UQL, etc.

**Comment by Jbalkan** [comment]
+1

**Comment by EP3UV2KyFG** [comment]
+1

**Comment by AbnormalRocks** [comment]
+1

**Comment by rmedeiros** [comment]
+1

**Comment by soussif** [comment]
+1

**Comment by imbananko** [comment]
+1

**Comment by fanmengwen** [comment]
+1

**Comment by duane-impulse** [comment]
+1

**Comment by murtazaqa** [comment]
+1

**Comment by giangduong** [comment]
+1

**Comment by yusufgungor** [comment]
+1

**Comment by ehsannm** [comment]
+1

**Comment by Hooooki** [comment]
+1

**Comment by shkarface** [comment]
+1

**Comment by bfoxstudio** [comment]
+1

**Comment by shubhhq** [comment]
+1

**Comment by jpascal** [comment]
+1

**Comment by robtaylorsky** [comment]
+1

**Comment by llybood** [comment]
+1

**Comment by TerralyticsNL** [comment]
+1

**Comment by Andrew713** [comment]
+1

**Comment by MatanyaLoewenthal** [comment]
Ok, like plus 1,000,000

**Comment by cenginner** [comment]
+1

**Comment by ajay-sainy** [comment]
+1

**Comment by Nurtau** [comment]
+1

**Comment by frederikmartin1337** [comment]
+1

**Comment by Multiply** [comment]
I was investigating library panels, and being able to reuse panels across multiple dashboards seems to work fine with variables, but you can't easily reuse the same library panel on the same dashboard, because you can't define panel-specific variables.

Specifically for library panels, it would make sense to be able to define variables that can either automatically be forwarded, or mapped directly on the panel.

**Comment by TouwaStar** [comment]
+1

**Comment by VermaPriyanka** [comment]
It will also drive improved dashboard performance with panel or row level filters, if a panel/row level filter value change  only re-renders the specific panel/row instead of the whole dashboard/all panels.

**Comment by yoni-dor-lmnd** [comment]
+1

**Comment by VikaChebotar** [comment]
+1

**Comment by ddarth** [comment]
+1

**Comment by dernat71** [comment]
@natellium I just saw that this request moved to `Backlog`. My team is interested in this feature. Can we assume it's on the roadmap somewhere?

**Comment by jad-gha** [comment]
My brothers and sisters in Christ just use this plugin : https://grafana.com/grafana/plugins/volkovlabs-variable-panel/

**Comment by hanssto** [comment]
This would be useful for default values for variables used in library panels. I want to make a library panel that supports filtering a metric successively on e.g. labels. But not all users will care about the drill-down.

Instead of forcing the dashboard owner to add a bunch of variables and default values to support my panel, they should add only those they care about. I'd set `label=~".*"` on the label matching as a default on the library panel level.

**Comment by xiepuhuan** [comment]
+1.I really need it

**Comment by stchenlucas** [comment]
> +1.I really need it

can you fix it

**Comment by xiepuhuan** [comment]
receive

**Comment by xiepuhuan** [comment]
> > +1.I really need it
> 
> can you fix it 你能解决吗

receive

**Comment by yossarin** [comment]
+1

**Comment by gilad-aperio** [comment]
+1, would unlock further value out of library panels.

**Comment by golan118a** [comment]
would be really useful and i have no idea how we got to this day without this feature 
+1

**Comment by eaugene** [comment]
+1 to this ask. This would be a great addition to have. particularly when the top panel shows the overall stats and the upcoming panels show the granular stats

**Comment by ewan-whiteswan** [comment]
+1

**Comment by ratbekn** [comment]
+1

**Comment by Etsija** [comment]
+1 would be very useful to us.

**Comment by danil-smirnov** [comment]
+1

**Comment by anhpham1509** [comment]
+1

**Comment by Kraego** [comment]
+1

**Comment by djamelb-np** [comment]
+1

**Comment by thomassjkim** [comment]
+1

**Comment by anatoly-kryzhanovsky** [comment]
+1

**Comment by vsafonin** [comment]
+1

**Comment by Max-E-Million** [comment]
+1

**Comment by fadedlamp42** [comment]
+1

**Comment by saikrishna-koppula** [comment]
+1

**Comment by Liquidmasl** [comment]
+1

**Comment by lukar-msft** [comment]
+1

**Comment by emerick-biron** [comment]
+1

**Comment by dariogmori** [comment]
+1

**Comment by mikhail-barg** [comment]
+1

**Comment by pogliad** [comment]
+1 Contra spem spero!

**Comment by Nordsoft91** [comment]
+1

**Comment by c-solo** [comment]
+1

**Comment by mbUSC** [comment]
+1

**Comment by yaelleC** [comment]
We just shipped variables for rows & tabs: https://grafana.com/whats-new/2026-04-08-stop-juggling-dashboards-with-grouping-level-variables/ 
I believe that would solve this feature request as you could create a row with a single panel?

**Comment by dmpe** [comment]
> https://grafana.com/whats-new/2026-04-08-stop-juggling-dashboards-with-grouping-level-variables/

https://grafana.com/whats-new/2026-04-08-stop-juggling-dashboards-with-variables-for-rows-and-tabs/

**Comment by hanssto** [comment]
My use case is library panels used for drilldowns. They might support tons of variables to be flexible for the user. Like "cluster", "namespace", "deployment", "pod", "container", "custom label X", etc. But then the dashboard has to define all the vars used in each panel, often as constants with default values like ".*", or you'll see "No data". One of the powers of the library panel is that people can customize it. It's hard for the user to inspect the panel to understand all the vars needed, and you'll violate DRY by repeating all the (unrelated) vars the panel can potentially use.

There are obviously counter-arguments regarding different dashboard design, using ad-hoc filters, and so on, but it gets very tricky when you have (promQL) joins and such.
